be/src/agent/publish_version.cpp

```diff
@@ -0,0 +1,161 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include "publish_version.h"
+
+#include <bvar/bvar.h>
+
+#include "gutil/strings/join.h"
+#include "storage/data_dir.h"
+#include "storage/storage_engine.h"
+#include "storage/tablet.h"
+#include "util/monotime.h"
+#include "util/threadpool.h"
+
+namespace starrocks {
+
+const uint32_t PUBLISH_VERSION_SUBMIT_MAX_RETRY = 10;
+bvar::LatencyRecorder g_publish_latency("be", "publish");
+
+void flush_dirs(std::unordered_set<DataDir*>& affected_dirs) {
```

Comment:
How about keeping using `txn_manager::persist_tablet_related_txns`?

Reply:
There are too many tablets, and you need to keep tabletsharedptrs references and iterate through them. On the other hand only keeping DataDir pointer is more lightweight, there are only a few(or just 1) datadirs.

Reply:
Flushing meta is a concept of `txn_manager`. Maybe a simplified version of `persist_tablet_related_txns` is what we need here. Otherwise, we would keep both `persist_tablet_related_txns` and `flush_dirs`.

Reply:
moved `flush_dirs` to txn_manager